superduperdb/ext/unstructured/encoder.py

```diff
@@ -0,0 +1,66 @@
+import dataclasses as dc
+import pickle
+import typing as t
+
+from unstructured.documents.elements import Element
+from unstructured.partition.auto import partition
+from unstructured.partition.html import partition_html
+
+from superduperdb.components.encoder import Encoder
+
+
+def link2element(link, unstructure_kwargs):
+    if link.startswith("file://"):
+        link = link[7:]
+
+    if link.startswith("http"):
+        # Special handling is required for http
+        elements = partition_html(url=link, **unstructure_kwargs)
+    else:
+        elements = partition(link, **unstructure_kwargs)
+    return elements
+
+
+def create_encoder(unstructure_kwargs):
+    def encoder(x: t.Union[str, t.List[Element]]):
+        if isinstance(x, str):
+            elements = link2element(x, unstructure_kwargs)
+        elif isinstance(x, list) and isinstance(x[0], Element):
+            elements = x
+        else:
+            raise TypeError(f"Cannot encode {type(x)}, must be str or list of Element")
+        return pickle.dumps(elements)
+
+    return encoder
+
+
+def create_decoder(unstructure_kwargs):
+    def decoder(b: bytes):
+        if isinstance(b, str):
+            return link2element(b, unstructure_kwargs)
+        try:
+            return pickle.loads(b)
+        except Exception as e:
+            # TODO: A compatibility issue with unstructured when using uri
+            # When superduperdb download uri and pass the bytes to the decoder,
+            # the file type message is lost, unstructured cannot  automatically parse.
+            raise ValueError(
+                "Cannot parse the bytes from uri data, please use encoder(x=uri)"
+            ) from e
+
+    return decoder
+
+
+@dc.dataclass
+class UnstructuredEncoder(Encoder):
+    unstructure_kwargs: t.Dict[str, t.Any] = dc.field(default_factory=dict)
+
+    def __post_init__(self):
+        self.encoder = create_encoder(self.unstructure_kwargs)
```

Comment:
What do these do? Convert a `.pdf` file to `bytes` and back?

Reply:
Here you want to keep all source data/`element`(page_num, text_type(title, text, table...)).
 If we don’t want the bytes, we can convert them into a `list of dict`, also can directly merge them into a text.
 
 Do you have any suggestion? 
 
 1. Add a tag to choose: `element`\`json`\`text`
 2. Add a merge func?
 3. only keep the source `element`

Reply:
Yes keeping all of that information is good.